File: src/lib/pjecalc/core/dominio/inss/faixas/faixa-previdenciaria.ts
/**
 * PJe-Calc v2.15.1 — FaixaPrevidenciaria
 * Porte 1:1 de: br.jus.trt8.pjecalc.negocio.dominio.inss.faixas.FaixaPrevidenciaria
 *
 * Ref Java: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/inss/faixas/FaixaPrevidenciaria.java
 *
 * Representa uma faixa do INSS progressivo:
 *   { valorInicial, valorFinal, aliquota }
 *
 * O cálculo `calcularValorNaFaixa(base)` retorna a contribuição DENTRO desta faixa:
 *   - Se base < valorInicial: 0
 *   - Senão: (base - valorInicial + 0.01) × (aliquota/100)
 *
 * Esse algoritmo permite somar contribuições de várias faixas para obter o INSS
 * total progressivo (usado desde EC 103/2019, mar/2020).
 */
import Decimal from 'decimal.js';
import { naoNulo, naoNulos, nulo, multiplicar, dividir, subtrair, CEM } from '../../../base/comum/utils';

const UM_CENTAVO = new Decimal('0.01');

export abstract class FaixaPrevidenciaria {
  protected valorInicial: Decimal | null;
  protected valorFinal: Decimal | null;
  protected aliquota: Decimal | null;

  constructor(valorInicial?: Decimal, valorFinal?: Decimal | null, aliquota?: Decimal) {
    this.valorInicial = valorInicial ?? null;
    this.valorFinal = valorFinal ?? null;
    this.aliquota = aliquota ?? null;
  }

  abstract getDiscriminador(): string;

  isValorFinalMenorOuIgualQueValorInicial(): boolean {
    if (!this.valorFinal || !this.valorInicial) return false;
    return this.valorFinal.comparedTo(this.valorInicial) <= 0;
  }

  /** adicionarNovoValorInicial (linha 62) — valorInicial = faixaAnterior.valorFinal + 0.01 */
  adicionarNovoValorInicial(faixa: FaixaPrevidenciaria | null): void {
    if (naoNulo(faixa) && faixa!.getValorFinal()) {
      this.valorInicial = faixa!.getValorFinal()!.plus(UM_CENTAVO);
    }
  }

  /** getValorMaximoDaFaixa (linha 68) */
  getValorMaximoDaFaixa(): Decimal | null {
    if (nulo(this.valorFinal)) return null;
    return multiplicar(
      subtrair(this.valorFinal, subtrair(this.valorInicial, UM_CENTAVO)),
      dividir(this.aliquota, CEM)
    );
  }

  /**
   * calcularValorNaFaixa (linha 75) — retorna a contribuição DENTRO desta faixa:
   *   - 0 se valorBase < valorInicial
   *   - (valorBase - valorInicial + 0.01) × (aliquota/100) caso contrário
   *
   * NOTA: para INSS progressivo, o caller deve limitar valorBase ao valorFinal
   * da faixa antes de chamar (ou usar min(valorBase, valorFinal)).
   */
  calcularValorNaFaixa(valorBase: Decimal | null): Decimal | null {
    if (nulo(valorBase)) return null;
    if (valorBase!.comparedTo(this.valorInicial!) < 0) return new Decimal(0);
    const valorBaseNaFaixa = subtrair(valorBase, subtrair(this.valorInicial, UM_CENTAVO))!;
    return multiplicar(valorBaseNaFaixa, dividir(this.aliquota, CEM));
  }

  sugerirValorMaximoParaValorFinal(): void { this.valorFinal = null; }

  getValorInicial(): Decimal | null { return this.valorInicial; }
  setValorInicial(v: Decimal): void { this.valorInicial = v; }
  getValorFinal(): Decimal | null { return this.valorFinal; }
  setValorFinal(v: Decimal | null): void { this.valorFinal = v; }
  getAliquota(): Decimal | null { return this.aliquota; }
  setAliquota(v: Decimal): void { this.aliquota = v; }
}

// Implementações concretas (1ª, 2ª, 3ª, 4ª, 5ª faixas)
// Cada uma só difere pelo discriminador (usado em mensagens de validação).

export class PrimeiraFaixaPrevidenciaria extends FaixaPrevidenciaria {
  getDiscriminador(): string { return '1'; }
}
export class SegundaFaixaPrevidenciaria extends FaixaPrevidenciaria {
  getDiscriminador(): string { return '2'; }
}
export class TerceiraFaixaPrevidenciaria extends FaixaPrevidenciaria {
  getDiscriminador(): string { return '3'; }
}
export class QuartaFaixaPrevidenciaria extends FaixaPrevidenciaria {
  getDiscriminador(): string { return '4'; }
}
export class QuintaFaixaPrevidenciaria extends FaixaPrevidenciaria {
  getDiscriminador(): string { return '5'; }
}

/**
 * Utilitário — calcula INSS progressivo total somando a contribuição de cada faixa.
 * Implementa o algoritmo:
 *   INSS = Σ faixa.calcularValorNaFaixa( min(base, faixa.valorFinal) )
 *
 * @param valorBase salário do mês
 * @param faixas    lista ordenada de faixas (valorInicial crescente)
 * @returns INSS total (não arredondado — caller aplica arredondamento HALF_EVEN)
 */
export function calcularInssProgressivo(
  valorBase: Decimal,
  faixas: FaixaPrevidenciaria[]
): Decimal {
  let total = new Decimal(0);
  for (const faixa of faixas) {
    const valorFinal = faixa.getValorFinal();
    const baseCapped = valorFinal !== null && valorBase.comparedTo(valorFinal) > 0 ? valorFinal : valorBase;
    const contribuicaoNaFaixa = faixa.calcularValorNaFaixa(baseCapped);
    if (contribuicaoNaFaixa !== null) {
      total = total.plus(contribuicaoNaFaixa);
    }
    // Se base menor que valorFinal, as próximas faixas não contribuem
    if (valorFinal !== null && valorBase.comparedTo(valorFinal) <= 0) break;
  }
  return total;
}

File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/inss/faixas/FaixaPrevidenciaria.java
/*
 * Decompiled with CFR 0.152.
 * 
 * Could not load the following classes:
 *  javax.persistence.Column
 *  javax.persistence.MappedSuperclass
 */
package br.jus.trt8.pjecalc.negocio.dominio.inss.faixas;

import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.negocio.comum.MensagemDeRecurso;
import br.jus.trt8.pjecalc.negocio.comum.exceptions.NegocioException;
import br.jus.trt8.pjecalc.negocio.constantes.Mensagens;
import java.io.Serializable;
import java.math.BigDecimal;
import javax.persistence.Column;
import javax.persistence.MappedSuperclass;

@MappedSuperclass
public abstract class FaixaPrevidenciaria
implements Serializable {
    private static final BigDecimal UM_CENTAVO = new BigDecimal("0.01");
    private static final long serialVersionUID = -1487119367940986877L;
    @Column
    private BigDecimal valorInicial;
    @Column
    private BigDecimal valorFinal;
    @Column
    private BigDecimal aliquota;

    public FaixaPrevidenciaria() {
    }

    public FaixaPrevidenciaria(BigDecimal valorInicial, BigDecimal valorFinal, BigDecimal aliquota) {
        this.valorInicial = valorInicial;
        this.valorFinal = valorFinal;
        this.aliquota = aliquota;
    }

    public FaixaPrevidenciaria validar(String nomeTabela, NegocioException excecao, boolean validarValorFinal) {
        if (Utils.nulo(this.getValorInicial())) {
            excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("valorInicialFaixa" + this.getDiscriminador() + nomeTabela, Mensagens.MSG0003, "Valor Inicial da Faixa " + this.getDiscriminador()));
        }
        if (validarValorFinal && Utils.nulo(this.getValorFinal())) {
            excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("valorFinalFaixa" + this.getDiscriminador() + nomeTabela, Mensagens.MSG0003, "Valor Final da Faixa " + this.getDiscriminador()));
        }
        if (Utils.nulo(this.getAliquota())) {
            excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("aliquotaFaixa" + this.getDiscriminador() + nomeTabela, Mensagens.MSG0003, "Al\u00edquota da Faixa " + this.getDiscriminador()));
        }
        if (Utils.naoNulos(this.getValorFinal(), this.getValorInicial()) && this.isValorFinalMenorOuIgualQueValorInicial()) {
            excecao.adicionarMensagemDeRecurso(new MensagemDeRecurso("valorFinalFaixa" + this.getDiscriminador() + nomeTabela, Mensagens.MSG0007, "Valor Final", "Valor Inicial"));
        }
        return this;
    }

    public abstract String getDiscriminador();

    public boolean isValorFinalMenorOuIgualQueValorInicial() {
        return this.valorFinal.compareTo(this.valorInicial) <= 0;
    }

    public void adicionarNovoValorInicial(FaixaPrevidenciaria faixa) {
        if (Utils.naoNulo(faixa)) {
            this.valorInicial = faixa.getValorFinal().add(UM_CENTAVO, Utils.CONTEXTO_MATEMATICO);
        }
    }

    public BigDecimal getValorMaximoDaFaixa() {
        if (this.valorFinal == null) {
            return null;
        }
        return Utils.multiplicar(Utils.subtrair(this.valorFinal, Utils.subtrair(this.valorInicial, UM_CENTAVO)), Utils.dividir(this.aliquota, Utils.CEM));
    }

    public BigDecimal calcularValorNaFaixa(BigDecimal valorBase) {
        if (valorBase == null) {
            return null;
        }
        if (valorBase.compareTo(this.valorInicial) < 0) {
            return BigDecimal.ZERO;
        }
        BigDecimal valorBaseNaFaixa = Utils.subtrair(valorBase, Utils.subtrair(this.valorInicial, UM_CENTAVO));
        return Utils.multiplicar(valorBaseNaFaixa, Utils.dividir(this.aliquota, Utils.CEM));
    }

    public void sugerirValorMaximoParaValorFinal() {
        this.valorFinal = null;
    }

    public BigDecimal getValorInicial() {
        return this.valorInicial;
    }

    public void setValorInicial(BigDecimal valorInicial) {
        this.valorInicial = valorInicial;
    }

    public BigDecimal getValorFinal() {
        return this.valorFinal;
    }

    public void setValorFinal(BigDecimal valorFinal) {
        this.valorFinal = valorFinal;
    }

    public BigDecimal getAliquota() {
        return this.aliquota;
    }

    public void setAliquota(BigDecimal aliquota) {
        this.aliquota = aliquota;
    }
}

